### DISCUSSION TITLE
Switch between CurrentHumidifierDehumidifierState using switch/case

### DISCUSSION THREAD

**Discussion by bramboe** [HomeSpan/HomeSpan#366] [Q&A] [answered]
He!

I'm working on creating a Dehumidifier accessoire. I'm trying to use the switch option by creating 3 'case' -states. In my logic I have to match the target case reference eg. case 0, case 1 etc.. with the Values of Dehumidifier States. The numbers don't seem to match, is this maybe due to the uint8_t-format? If so, how do I get the right values?

My setup ( I left some code out  )

```
struct DEV_HumSensor : Service::HumidifierDehumidifier {     
  SpanCharacteristic *currentState;
  SpanCharacteristic *targetState;


  DEV_HumSensor(int dhtPin, int buttonPin, int writePin) : Service::HumidifierDehumidifier() {      
    power=new Characteristic::Active();
    currentState = new Characteristic::CurrentHumidifierDehumidifierState(2);

    (targetState = new Characteristic::TargetHumidifierDehumidifierState())->setValidValues(3,0,2,3);         //removed humidifier state as it does not exist.           
}
```

```
boolean update(){     

int state = targetState->getNewVal();
    switch (state) {
      case 0: 
        LOG1("Case0");
        LOG1(state);
        break;
     case 2: 
     --  etc -- 

    return(true);    
```

**Comment by HomeSpan** [comment]
What numbers are you seeing?  Your *state* variable should be returning the right levels.  Note that the the TARGET and CURRENT Characteristics do not match one-for-one.  They represent slightly different things.

**Comment by bramboe** [comment]
Morning!

When I press AUTO

```
EVENT/1.0 200 OK
Content-Type: application/hap+json
Content-Length: 49

{"characteristics":[{"iid":9,"value":2,"aid":1}]}
-------- SENT ENCRYPTED! --------
Temperature Update: 25.00 ; 
>>>>>>>>>> 192.168.178.109 >>>>>>>>>>
EVENT/1.0 200 OK
Content-Type: application/hap+json
Content-Length: 51

```

When I press DEHUMIDIFY
```
Content-Length: 78
Content-Type: application/hap+json
------------ END BODY! ------------
{"characteristics":[{"aid":1,"iid":8,"value":1},{"aid":1,"iid":12,"value":2}]}
------------ END JSON! ------------
In Put Characteristics #2 (192.168.178.109)...
:  Current Power=true  New Power=true
Case2
Read state: 2
Updating aid=1 iid=8 (okay)
Updating aid=1 iid=12 (okay)
```

When I press OFF
```
{"characteristics":[{"aid":1,"iid":8,"value":0}]}
------------ END JSON! ------------
In Put Characteristics #2 (192.168.178.109)...
:  Current Power=true  New Power=false
Case2
Read state: 2
Requested switch OFF the device.
Updating aid=1 iid=8 (okay)
```

**Comment by HomeSpan** [comment, marked as answer]
Note that there are two separate Characteristics.  One for Active and one for the Target HumidifierDehumidifier State.  The later does not have an "off" setting - it's either Auto, Humidify, or Dehumidify.  When you press "DeHumidify" in the Home App you get TWO new updates - one for IID=8 and one for IID=12, which will match your NewPower=true and Case2.  But when you press "OFF", the only thing the HomeApp transmits is an update the the Active Characteristic, which is IID=8, and correctly shows NewPower=false.  However there is no change in the Target State (no update to IID=12) sent.  Your code however is running the Switch statement even though Target State has not been updated.  There is nothing wrong with that, but note that it will simply echo the latest value of Target State, which is still equal to 2 (Dehumidify). 

You can check to see if Target State was update using the `isUpdated()` method, and skip your Switch statement if needed.

Why did Apple make this so complicated?  The reason is that you can also control the Humidifier by simply pressing the Accessory Tile in the Home App.  This "activates" the Humidifier and causes the Home App to send an update to IID=8, setting Active to either TRUE or FALSE.  There is no change to the Target State - it remains the same as whatever it was when the Humidifier was last turned off.

Within your code, to determine whether to turn on/off the humidifier, you'll therefore need to look at BOTH the Active and TargetState Characteristics, determine which ones were update by the Home App, and thus what actions to take, if any.

**Reply by bramboe** [reply]
Thank you for the explanation on how the Active & states work. That clarifies a lot :) 
Based on what you said regarding 'isUpdated()', I'm not familiar with that method. I've been searching for references within Homespan to understand more - besides the comment in 09-MessageLogging and usage (boolean isUpdated) in HomeSpan.h - can you maybe point me in the right direction?

**Reply by HomeSpan** [reply]
Sorry, I provided the wrong reference.  The method is just updated(), not isUpdated().  You'll find the docs for it under the SpanCharacteristic section of the API Reference.

**Comment by bramboe** [comment]
Ok, I mapped the cases, I can add actions to the right buttons. Somehow I can't seem to turn on the device when I press Dehumidify or Auto. I have to press the Off button (power->getNewVal())) to turn it off and on. Any idea why this happens?

I have 2 void's to turn and off the device. 

```
  void powerOn() {
    LOG1("Requested switch ON the device.\n");
    power->setVal(1);
    digitalWrite(this->writePin, HIGH);

  }
  void powerOff() {
    LOG1("Requested switch OFF the device.\n");
    power->setVal(0);
    digitalWrite(this->writePin, LOW);

  }
```

I do see the LOG "Requested switch ON the device." when I press auto or dehumidify but nothing happens.


```
boolean update(){                              // update() method

int state = targetState->getNewVal();
    switch (state) {
      case 0: //AUTO
            LOG1("Auto \n");
            targetState->setVal(0); 
            currentState->setVal(0);

        break;
     case 2: // DEHUMIDIFY
            LOG1("Dehumidify \n");
            targetState->setVal(2);
            currentState->setVal(2);
        break;

    }  


    if (power->getNewVal()==0){
      this->powerOff();
    } 
    else {
      this->powerOn();
    }

    return(true);                               // return true
```

**Comment by HomeSpan** [comment]
I'm not sure I fully understand the issue - are you saying the log file shows you reach powerOn() but the digitalWrite() is not turning on the physical device?   Or is powerOn() simply not reached unless you press ON on in the Home App

One thing I noted is that you are using setVal() to change the values of targetState and power.  These should be removed.  The Home App is setting this for you - that's how you know what actions are being taken by the Home App.  The purpose of setVal is to update the other Characteristics, such as currentState, that are not changed by the Home App but are instead read back by the Home App to set its own status.

The only time you would use setVal to change power or targetState is if you have physical buttons on your device that would turn on the appliance without the Home App.  In this case you would use setVal to change, for example power, to tell the Home App the state of power has indeed changed (for a reason not associated with the Home App).

**Reply by bramboe** [reply]
First of all, thank you. Offtopic; I wanted to support you but didn't see a button on your project page!

To answer your question: yes, If the device is turned off, and I press auto or dehumidify (Homekit) I'm seeing the LOG Requested switch ON the device -but nothing happens. When I press the 'off' button (homekit) it will turn on and if pressing the 'off' button again (homekit) it turns off.

The ESP32 is connected to the physical button, but I've only been focusing on the homekit setup sending a HIGH-digitalwrite (pin on output) to the button, and not reading the button when it's physically pressed:) 

My log when pressing auto, dehumidify & off in homekit:

Pressed Dehumidify
```
Dehumidify 
Requested switch ON the device.
Power State:1
Target State:2
Target State updated:1
Updating aid=1 iid=12 (okay)
Updating aid=1 iid=8 (okay)
```

Pressed auto:
```
Auto 
Requested switch ON the device.
Power State:1
Target State:0
Target State updated:1
Updating aid=1 iid=8 (okay)
Updating aid=1 iid=12 (okay)
```

Pressed OFF:
```
Auto 
Requested switch OFF the device.
Power State:0
Target State:0
Target State updated:0
Updating aid=1 iid=8 (okay)
```

**Comment by HomeSpan** [comment]
It sounds like this is an issue with how your electrical circuit is constructed.   How is the ESP32 connected to the humidifier?  Based on your code it would need to be through a relay or or sensor that is active when the pin is HIGH and inactive when the pin is LOW.  But there are many other relays and combinations that are possible, including latch-relays, relays that trigger on a HIGH pulse, LOW pulse, etc.  That could explain some of what you are seeing.

Re donations, thanks for the offer, but I'm just glad others are finding the library useful and appreciate all the feedback and great suggestions for improving HomeSpan.

**Reply by bramboe** [reply]
Yes, I think you're right. If I use a simple switch on/off it seems to work just fine. But with the toggle options of auto, dehumidify and off, it only responds to 'OFF', pressing back and fort between the 3 options, the 'OFF' will also turn it back on. So it has something to do when the pin is LOW. I have the ESP32 GPIO5 & GND connected to both legs of the microswitch (on an existing PCB board.)

**Comment by bramboe** [comment]
I think I got it to work, so here I'm sharing. The ultimate goal is to control the device with all its functions. I'm starting with the basic options: off, dehumidify & auto. Later I will connect the 'auto' to another option on the device. Next steps is to update Homekit when the controls are physically pressed. 

Thank you for all your input and helping me understand :)

```

boolean update(){                              // update() method
if(targetState->updated() == 0 && power->updated() == 0) {
    power->setVal(0);
    digitalWrite(this->writePin, 1);
    delay(250);    
    digitalWrite(this->writePin, 0);
    delay(250);       
    } else {

  int state = targetState->getNewVal();
      switch (state) {
        case 0: //AUTO
        LOG1("Auto \n");
        currentState->setVal(0);
  if (power->getVal() == 0) {
      power->setVal(1);
      digitalWrite(this->writePin, 1);
      delay(250);    
      digitalWrite(this->writePin, 0);
      delay(250);      
      }
 
          break;
       case 2: // DEHUMIDIFY
       LOG1("Dehumidify \n");
       currentState->setVal(2);
  if (power->getVal() == 0) {
      power->setVal(1);
      digitalWrite(this->writePin, 1);
      delay(250);    
      digitalWrite(this->writePin, 0);
      delay(250);    
  
  
  
      }

          break;
  
      }
}

    return(true);                               // return true
    
  } // update
```
